### ISSUE TITLE
killing a background process?

### ISSUE THREAD

**Issue by mitchellwrosen** [Gabriel439/Haskell-Turtle-Library#103]
I'm writing some acceptance tests that require spawning a process that runs forever under normal circumstances, and I'd like to kill it between tests and such. Is there any way to do that from within `turtle`? It looks like `shell`, `proc` and friends all call `System.Process.waitForProcess`, but don't give you access to the underlying `ProcessHandle` to send a SIGINT with `System.Process.terminateProcess`

**Comment by Gabriel439** [comment]
The simplest workaround without exposing `process` internals is to fork a thread and kill that:

```haskell
example = do
    async <- fork (proc ...)
    ...
    cancel async

**Comment by mitchellwrosen** [comment]
Correct me if I'm wrong, but wouldn't that only kill the thread waiting on the process to return, and not the actual process itself?

**Comment by Gabriel439** [comment]
You're right.  The implementation of `proc` internally uses a function called `system` which `bracket`s the process call to ensure proper `Handle` shutdown, but doesn't actually cancel the process.  I could modify it to cancel the process, too, assuming that doesn't break anything else.

**Comment by mitchellwrosen** [comment]
It would break code that relies on the process still being alive after `cancel`ing its associated async :)

**Comment by mitchellwrosen** [comment]
Well, I'm in favor of this change, since I need it :P

If you're on board I can send a PR

**Comment by Gabriel439** [comment]
I'm okay with breaking any code that assumes that.  I feel like distinguishing the process thread from the surrounding thread is a misfeature.  From the user's perspective the process should appear like a synchronous `IO` action within the same thread.

Having to keep track of one extra thread is one more moving part that can break in surprising ways.

**Comment by mitchellwrosen** [comment]
I hear you, but this change would only affect `fork`ed processes, which should feel like _asynchronous_ IO actions. Right?

**Comment by Gabriel439** [comment]
Yeah, `system` already waits for the process to complete anyway, so we should complete the illusion by translating all thread kills to process kills, too.

**Comment by mitchellwrosen** [comment]
```haskell
system
    :: Process.CreateProcess
    -- ^ Command
    -> Shell Text
    -- ^ Lines of standard input
    -> IO ExitCode
    -- ^ Exit code
system p s = do
    let p' = p
            { Process.std_in  = Process.CreatePipe
            , Process.std_out = Process.Inherit
            , Process.std_err = Process.Inherit
            }
    liftIO $ bracket 
        (Process.createProcess p')
        (\(_, _, _, ph) -> Process.terminateProcess ph) $
        \(Just hIn, Nothing, Nothing, ph) -> do
            let feedIn = sh (do
                txt <- s
                liftIO (Text.hPutStrLn hIn txt) )
            withAsync feedIn (\_ -> liftIO (Process.waitForProcess ph) )
```
?

**Comment by Gabriel439** [comment]
Fixed by #104

**Comment by Gabriel439** [comment]
Thanks for the pull request! :)

**Comment by mitchellwrosen** [comment]
Heh, I've hit another sort of annoying problem with this - I've forked a thread that's sitting in a blocking FFI call forever. The only thing that can cause it to die is by signaling SIGINT or SIGTERM. `cancel` and `waitAsync` and friends cause an indefinite block, as `throwTo` won't return until Haskell regains control:

    If the target thread is currently making a foreign call, then the exception will not be raised (and hence    
    throwTo will not return) until the call has completed. This is the case regardless of whether the call is 
    inside a mask or not. However, in GHC a foreign call can be annotated as interruptible, in which case 
    a throwTo will cause the RTS to attempt to cause the call to return; see the GHC documentation for 
    more details.

I'd like to read the mentioned GHC docs, but I'm not sure what section this is referring to. Anywho, if I can get the forked process's `ProcessHandle`, I can just call `terminateProcess` myself. But, this is starting to feel a bit out of scope for `turtle`. Thoughts?

**Comment by Gabriel439** [comment]
You can also make your FFI call non-blocking by annotating it as `safe`, if you are willing to pay a small performance overhead.

**Comment by mitchellwrosen** [comment]
That won't make it non-blocking wrt asynchronous exceptions. `interruptible`, however, will :) I just made a PR to `inline-c` to let me do this.

**Comment by capitanbatata** [comment]
I'm also stuck with the same issue (or a similar one). I'm running an infinite process (a Java program) and I cannot cancel it using [race](https://hackage.haskell.org/package/async-2.1.0/docs/Control-Concurrent-Async.html#v:race):

```haskell
    ...
    res <- liftIO $ sutProc `race` testerProc
    ...
    where
      sutProc = sh $ using $ do
        a <- fork (proc "java" ["SomeClass"] mempty)
        Turtle.wait a
```

Where the `testerProc` terminates after performing a couple of tests.

I thought @mitchellwrosen fixed this issue when using `fork`. Maybe I'm not using it in the right way. Any hints?

**Comment by Gabriel439** [comment]
@capitanbatata: What happens if you change `sutProc` to:

```haskell
sutProc = proc "java" ["SomeClass"] mempty
```

**Comment by capitanbatata** [comment]
That was the (very elegant code ;)) I had before. In that case `race` (and in turn `cancel`) won't cancel the process, since it involves an external call and (according to what I could understand from the documentation) it is waiting for the process (that involves an infinite loop in my case) to terminate.

What does work is something like:

```haskell
...
    res <- liftIO $ sutProc `race` testerProc
    ...
    where
      sutProc = ph <- Process.spawnProcess "/usr/bin/java" ["SomeClass"]
      (forever $ sh (sleep 2.0) >> print "Waiting for process") `onException` Process.terminateProcess ph     
```

But is quite ugly, does not make use of the full power of Turtle, and forces me to change a lot of types (from FilePath to String for instance). But is the only solution I could find so far.

Strangely enough, if instead of `spawnProcess` I use `spawnCommand` `terminateProcess` does not kill the Java program (even though it can be killed by `kill`).

**Comment by mitchellwrosen** [comment]
@capitanbatata Coincidentally, I just made a small patch to process that lets you `cancel` a `waitForProcess` (https://github.com/haskell/process/pull/101). Might that solve your issue?

**Comment by capitanbatata** [comment]
That looks like it might help @mitchellwrosen . I can't use the latest process with `stack` and `lts-9.0`. I might have to use a nightly build...

**Comment by capitanbatata** [comment]
BTW, does the above patch make cancellation of processes spawned by `Turtle`'s fork work? Or does this library needs to be patched as well?

**Comment by Gabriel439** [comment]
@capitanbatata: What happens if you replace the `mempty` input to `proc` with `empty` instead.  Note that they are not the same:

```haskell
Prelude Turtle> stdout empty  -- `empty` emits 0 line
Prelude Turtle> stdout mempty -- `mempty` emits a single empty line

Prelude Turtle>
```

Also, in general either `empty` or `stdin` is what you should be feeding to `proc`/`procs` invocations and usually not `mempty`

**Comment by capitanbatata** [comment]
Thanks for the clarification @Gabriel439. I didn't know about that! 

To help the discussion I extracted the gist of the problem into the following example:
    https://github.com/capitanbatata/sandbox/tree/master/racing-turtles

There you can see that I can cancel the processes started by `internalExample` and `sysProcExceptionExample`, but I cannot cancel the processes started by the Turtle functions. I don't know if there is a way to solve this using the functions provided by this library.

**Comment by Gabriel439** [comment]
`turtle`'s `fork` is just a thin wrapper around `withAsync`, so you can easily reproduce the problem without `turtle` by replacing this:

```haskell
sh $ using $ do
          a <- fork highlander
          Turtle.wait a
```

... with this:

```haskell
Control.Concurrent.Async.withAsync highlander Control.Concurrent.Async.wait
```

That doesn't mean that `turtle` is not to blame, but if it is the `fork` function is definitely not the problem (it could be an issue with `proc`, for example)

You can further narrow this down by trying to replace `Turtle.proc` with the use of the underlying `process` library

**Comment by capitanbatata** [comment]
> That doesn't mean that turtle is not to blame, but if it is the fork function is definitely not the problem (it could be an issue with proc, for example)
Ideally I'd like to use the `procExample` version, which uses `Turtle.proc`. But this requires that `proc` cancels the external process upon receiving an asynchronous exception. But I don't know if that's the intended behavior of `proc`. If so I could try to take a look at the implementation and adapt it. 

> You can further narrow this down by trying to replace Turtle.proc with the use of the underlying process library

This is exactly what I do in `sysProcExceptionExample` in the code I sent above. In that case I can cancel the process, but I have to call `Process.terminateProcess`. And this solution is extremely cumbersome, as you can see. That's why I was trying to use `turtle` :)

**Comment by capitanbatata** [comment]
I observe different behaviors on Windows and Linux in the example I sent, so I'm making an issue with the relevant details to stop polluting this thread.

**Comment by cideM** [comment]
I am seeing this issue as well

```haskell
module Main where

import Control.Concurrent
import Control.Concurrent.Async
import Control.Monad
import Control.Retry
import Data.ByteString (ByteString)
import qualified Data.Text.Encoding as T
import Network.HTTP.Req
import Test.Tasty
import Test.Tasty.HUnit
import Turtle

runServer :: IO ExitCode
runServer = proc "cabal" ["v2-run", "--cabal-file", "./backend/lions-backend.cabal", "run-lions-backend"] empty

httpOptions =
  defaultHttpConfig
    { httpConfigRetryJudgeException = \_ _ -> True,
      httpConfigRetryPolicy = constantDelay 500000 <> limitRetries 50
    }

waitForServer :: IO ()
waitForServer = do
  _ <- runReq httpOptions $ do
    result <- req GET (http "localhost" /: "login") NoReqBody bsResponse (port 3000)
    return (responseStatusCode result :: Int)
  pure ()

main :: IO ()
main = do
  sh $ do
    foo <- fork runServer

    liftIO $ waitForServer

    liftIO $
      defaultMain $ do
        testCase "Should show login page" $ do
          "" @?= ""

    liftIO $ cancel foo
```

Running this leaves `v2-run` running :(
